lib/railway_ipc/core/message_format/binary_protobuf.ex

```diff
@@ -0,0 +1,153 @@
+  message type as a string in colon format.
+
+  Encodes the given `protobuf` by creating a JSON string with two attributes:
+
+  * `type` -- the Protobuf module name as a string using colon notation
+  * `encoded_message` -- the Base64 encoded Protobuf
+
+  """
+  def encode(protobuf) when is_map(protobuf) or is_atom(protobuf) do
+    protobuf
+    |> build_payload()
+    |> encode_type()
+    |> encode_message()
+    |> encode_payload_as_json()
+  end
+
+  def encode(_), do: {:error, "Argument Error: Valid Protobuf required"}
+
+  @doc """
+  Decodes the given `message` into a Protobuf `struct`.
+
+  """
+  def decode(message) when not is_binary(message) do
+    {:error, "Malformed JSON given. Must be a string. (#{inspect(message)})"}
+  end
+
+  def decode(message) do
+    message
+    |> decode_json()
+    |> parse_type()
+    |> check_that_module_is_defined()
+    |> decode_protobuf()
+  end
+
+  defp build_payload(protobuf), do: {:ok, %{}, protobuf}
+
+  defp encode_type({:ok, payload, protobuf}) do
+    type =
+      protobuf.__struct__
+      |> to_string
+      |> replace_invalid_chars()
+
+    {:ok, Map.put(payload, :type, type), protobuf}
+  rescue
+    KeyError ->
+      {:error, "Argument Error: Valid Protobuf required"}
+  end
+
+  defp encode_message({:ok, payload, protobuf}) do
+    encoded_message =
+      protobuf
+      |> protobuf.__struct__.encode()
+      |> Base.encode64()
+
+    {:ok, Map.put(payload, :encoded_message, encoded_message), protobuf}
+  end
+
+  defp encode_message({:error, _} = error), do: error
+
+  defp encode_payload_as_json({:ok, payload, _}) do
+    case Jason.encode(payload) do
+      {:ok, json} -> {:ok, json, payload.type}
+      {:error, error} -> {:error, error}
+    end
+  end
+
+  defp encode_payload_as_json({:error, _} = error), do: error
+
+  defp parse_type({:ok, %{type: type} = payload}) when is_binary(type) do
+    {:ok, Map.put(payload, :module, type_to_module(type))}
+  end
+
+  defp parse_type({:ok, %{type: type}}) when not is_binary(type) do
+    {:error, "Message `type` attribute must be a string"}
+  end
+
+  defp parse_type({:ok, _}), do: {:error, "Message is missing the `type` attribute"}
+  defp parse_type({:error, _} = error), do: error
+
+  defp check_that_module_is_defined({:ok, payload}) do
+    %{module: module} = payload
+    module.__info__(:module)
+    {:ok, payload}
+  rescue
+    UndefinedFunctionError ->
+      {:unknown_message_type, Map.put(payload, :module, DefaultMessage)}
+  end
+
+  defp check_that_module_is_defined({:error, _} = error), do: error
+
+  defp decode_protobuf({:error, _} = error), do: error
+
+  defp decode_protobuf({status, payload}) do
+    %{module: module, encoded_message: encoded_message, type: type} = payload
+
+    decoded_message =
+      encoded_message
+      |> Base.decode64!(ignore: :whitespace)
+      |> module.decode
+
+    {status, decoded_message, type}
+  rescue
+    # TODO: What's the specific error we should rescue here? Can't find
```

Comment:
Did you ever determine this? I am not very familiar with the specific error myself though either.

Reply:
Not yet; punting for now since the original code didn't event handle this scenario